ds701_book/20-Recommender-Systems.qmd

```diff
@@ -642,87 +884,80 @@ Again, this is only on the training data -- so overly optimistic for sure.
 And note that this is a subset of the full dataset -- the subset that is "easiest" to predict due to density.
 
 ```{python}
-#| slideshow: {slide_type: fragment}
 from sklearn.neighbors import KNeighborsClassifier
 from sklearn.metrics import mean_squared_error
+
+# Drop the columns that are not features
 X_train = good_df.drop(columns=['Id', 'ProductId', 'UserId', 'Text', 'Summary'])
+
+# The target is the score
 y_train = good_df['Score']
+
 # Using k-NN on features HelpfulnessNumerator, HelpfulnessDenominator, Score, Time
 model = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train)
 %time y_hat = model.predict(X_train)
 ```
 
 ```{python}
-#| hide_input: true
-#| slideshow: {slide_type: fragment}
-#| tags: [remove-input]
 print(f'RMSE on visible entries (test set): {mean_squared_error(y_train, y_hat, squared = False):0.3f}')
 ```
 
-### Assessing MF
+## Assessing Matrix Factorization
 
-Matrix Factorization per se is a good idea.    However, many of the improvements we've discussed for CF apply to MF as well.
+Matrix Factorization per se is a good idea.    
+However, many of the improvements we've discussed for CF apply to MF as well.
 
-To illustrate, we'll look at some of the successive improvements used by the team that won the Netflix prize ("BellKor's Pragmatic Chaos").
+To illustrate, we'll look at some of the successive improvements used by the
+team that won the Netflix prize ("BellKor's Pragmatic Chaos").
 
 When the prize was announced, the Netflix supplied solution achieved an RMSE of 0.951.
 
 By the end of the competition (about 3 years), the winning team's solution achieved RMSE of 0.856.
 
 Let's restate our MF objective in a way that will make things clearer:
 
-$$ \min_{U, V} \sum_{(u, i)\in S}(r_{ui} - u_u^Tv_i)^2 + \lambda(\Vert U\Vert^2 + \Vert V\Vert^2) $$
+$$
+\min_{U, V} \sum_{(u, i)\in S}(r_{ui} - u_u^Tv_i)^2 + \lambda(\Vert U\Vert^2 + \Vert V\Vert^2) 
```

Comment:
I think we should mention why we now have a sum as opposed to a full matrix equation.